torch/nested/_internal/nested_tensor.py

```diff
@@ -44,27 +68,34 @@ def __new__(cls, values, offsets, **kwargs):
             False,
             ks,
         )
-        # TODO: why is values requires grad?
-        # if r.requires_grad:
-        #     raise ValueError(
-        #         "buffer should not require grad when constructing NestedTensor")
-        r._values = values.detach() if values.requires_grad else values
         return r
 
-    def __init__(self, values, offsets, **kwargs):
+    def __init__(self, values, offsets, *, ragged_size=None, **kwargs):
         super().__init__()
         # Only support jagged for now.
         assert offsets is not None
         assert offsets.ndim == 1
         assert not isinstance(values, NestedTensor)
         assert values.ndim == 2
 
-        # In a later PR, we'll need to accept an additional size argument
-        # to handle dynamic shapes.
-        ragged_dim = get_tensor_id(offsets, coeff=1)
+        if ragged_size is None:
```

Comment:
Will these two things always happen 1-to-1? (both true or both false)

(1) `ragged_size` is specified in the constructor
(2) `values` and `offsets` are fake tensors.

Maybe we should assert that if it's the case

Reply:
You're right in that it COULD be that way (as in its only useful if we need to propagate the symbolic ragged size), but that is not the case today just because it seemed easier to me to have the torch dispatch logic, in particular `extract_kwargs` to do the same thing during eager and during tracing: always propagate the ragged_size along if I'm passing the offsets along anyway.